From 485be336ae8fb947e5d426957dc9f60ff56045f2 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Persistent Meta API

This adds an API for getting and setting PERSISTENT Metadata
Meta data can be stored on Worlds, Entities, Chunks and at a Block level.
Meta is is provided in a simple Map/List setup and does not expose raw NBT access.
---
 .../metaapi/BlockWithMetaClearedEvent.java         |  36 ++
 .../java/com/empireminecraft/metaapi/MetaApi.java  | 381 +++++++++++++++++++
 src/main/java/net/minecraft/server/Chunk.java      |   1 +
 .../net/minecraft/server/ChunkRegionLoader.java    |   2 +
 src/main/java/net/minecraft/server/Entity.java     |   3 +
 .../java/net/minecraft/server/MetaApiAccessor.java | 420 +++++++++++++++++++++
 .../minecraft/server/PacketPlayOutMapChunk.java    |   1 +
 .../server/PacketPlayOutTileEntityData.java        |   1 +
 src/main/java/net/minecraft/server/TileEntity.java |   3 +
 src/main/java/net/minecraft/server/World.java      |   1 +
 src/main/java/net/minecraft/server/WorldData.java  |   5 +-
 11 files changed, 853 insertions(+), 1 deletion(-)
 create mode 100644 src/main/java/com/empireminecraft/metaapi/BlockWithMetaClearedEvent.java
 create mode 100644 src/main/java/com/empireminecraft/metaapi/MetaApi.java
 create mode 100644 src/main/java/net/minecraft/server/MetaApiAccessor.java

diff --git a/src/main/java/com/empireminecraft/metaapi/BlockWithMetaClearedEvent.java b/src/main/java/com/empireminecraft/metaapi/BlockWithMetaClearedEvent.java
new file mode 100644
index 0000000..4114b48
--- /dev/null
+++ b/src/main/java/com/empireminecraft/metaapi/BlockWithMetaClearedEvent.java
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2016. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ *
+ */
+
+package com.empireminecraft.metaapi;
+
+import net.minecraft.server.BlockPosition;
+
+import net.minecraft.server.Chunk;
+import org.bukkit.block.Block;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class BlockWithMetaClearedEvent extends Event {
+    private final Block block;
+    public BlockWithMetaClearedEvent(Chunk chunk, BlockPosition pos) {
+        this.block = chunk.bukkitChunk.getBlock(pos.getX(), pos.getY(), pos.getZ());
+    }
+
+    public Block getBlock() {
+        return block;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/metaapi/MetaApi.java b/src/main/java/com/empireminecraft/metaapi/MetaApi.java
new file mode 100644
index 0000000..19a1eb6
--- /dev/null
+++ b/src/main/java/com/empireminecraft/metaapi/MetaApi.java
@@ -0,0 +1,381 @@
+package com.empireminecraft.metaapi;
+
+import com.google.common.collect.ForwardingMap;
+import net.minecraft.server.MetaApiAccessor;
+import org.bukkit.Chunk;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.block.BlockState;
+import org.bukkit.craftbukkit.CraftChunk;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.entity.Entity;
+import org.bukkit.inventory.ItemStack;
+
+import java.security.InvalidParameterException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+@SuppressWarnings({"WeakerAccess", "unused"})
+public final class MetaApi {
+    private MetaApi() {}
+
+    //////////////////////////////////////////////////
+    //// WORLD META
+    //////////////////////////////////////////////////
+
+    public static MetaMap getWorldMetaMap(World world) {
+        return MetaApiAccessor.getWorldMetaMap((CraftWorld) world);
+    }
+
+    public static <T> T getWorldMeta(World world, String key) {
+        return getWorldMeta(world, key, null);
+    }
+
+    public static <T> T getWorldMeta(World world, String key, T def) {
+        final MetaMap worldMetaMap = getWorldMetaMap(world);
+        T ret = worldMetaMap != null && !worldMetaMap.isEmpty() ? (T) worldMetaMap.get(key) : null;
+        return ret != null ? ret : def;
+    }
+
+    public static boolean hasWorldMeta(World world, String key) {
+        final MetaMap worldMetaMap = getWorldMetaMap(world);
+        return worldMetaMap != null && !worldMetaMap.isEmpty() && worldMetaMap.containsKey(key);
+    }
+
+    public static <T> T setWorldMeta(World world, String key, Object val) {
+        return setMetaMapValue(getWorldMetaMap(world), key, val);
+    }
+
+    public static <T> T removeWorldMeta(World world, String key) {
+        return setWorldMeta(world, key, null);
+    }
+    
+    //////////////////////////////////////////////////
+    //// ENTITY META
+    //////////////////////////////////////////////////
+
+    public static MetaMap getEntityMetaMap(Entity entity, boolean isWrite) {
+        return MetaApiAccessor.getEntityMetaMap(entity, isWrite);
+    }
+
+    public static <T> T getEntityMeta(Entity entity, String key) {
+        return getEntityMeta(entity, key, null);
+    }
+
+    public static <T> T getEntityMeta(Entity entity, String key, T def) {
+        final MetaMap entityMetaMap = getEntityMetaMap(entity, false);
+        T ret = entityMetaMap != null && !entityMetaMap.isEmpty() ? (T) entityMetaMap.get(key) : null;
+        return ret != null ? ret : def;
+    }
+
+    public static boolean hasEntityMeta(Entity entity, String key) {
+        final MetaMap entityMetaMap = getEntityMetaMap(entity, false);
+        return entityMetaMap != null && !entityMetaMap.isEmpty() && entityMetaMap.containsKey(key);
+    }
+
+    public static <T> T setEntityMeta(Entity entity, String key, Object val) {
+        return setMetaMapValue(getEntityMetaMap(entity, true), key, val);
+    }
+
+    public static <T> T removeEntityMeta(Entity entity, String key) {
+        return setMetaMapValue(getEntityMetaMap(entity, false), key, null);
+    }
+
+    //////////////////////////////////////////////////
+    //// CHUNK META
+    //////////////////////////////////////////////////
+
+    /**
+     * Only provided for low level access as debug command
+     * @param chunk
+     * @return
+     */
+    @Deprecated
+    public static Map<BlockPos, MetaMap> getChunkMetaData(Chunk chunk) {
+        return ((CraftChunk) chunk).getHandle().chunkMeta;
+    }
+    public static MetaMap getChunkMetaMap(Chunk chunk, boolean isWrite) {
+        return MetaApiAccessor.getChunkMetaMap(chunk, isWrite);
+    }
+
+    public static boolean hasChunkMeta(Chunk chunk, String key) {
+        final MetaMap chunkMetaMap = getChunkMetaMap(chunk, false);
+        return chunkMetaMap != null && !chunkMetaMap.isEmpty() && chunkMetaMap.containsKey(key);
+    }
+
+    public static <T> T getChunkMeta(Chunk chunk, String key) {
+        return getChunkMeta(chunk, key, null);
+    }
+
+    public static <T> T getChunkMeta(Chunk chunk, String key, T def) {
+        final MetaMap chunkMetaMap = getChunkMetaMap(chunk, false);
+        T ret = chunkMetaMap != null && !chunkMetaMap.isEmpty() ? (T) getChunkMetaMap(chunk, false).get(key) : null;
+        return ret != null ? ret : def;
+    }
+
+    public static <T> T setChunkMeta(Chunk chunk, String key, Object val) {
+        return setMetaMapValue(getChunkMetaMap(chunk, true), key, val);
+    }
+
+    public static <T> T removeChunkMeta(Chunk chunk, String key) {
+        return setMetaMapValue(getChunkMetaMap(chunk, false), key, null);
+    }
+
+    //////////////////////////////////////////////////
+    //// BLOCK META
+    //////////////////////////////////////////////////
+
+    public static MetaMap getBlockMetaMap(Location loc, boolean isWrite) {
+        return MetaApiAccessor.getBlockMetaMap(loc, isWrite);
+    }
+
+    public static boolean hasBlockMeta(Location loc, String key) {
+        final MetaMap blockMetaMap = getBlockMetaMap(loc, false);
+        return blockMetaMap != null && !blockMetaMap.isEmpty() && blockMetaMap.containsKey(key);
+    }
+
+    public static <T> T getBlockMeta(Location loc, String key) {
+        return getBlockMeta(loc, key, null);
+    }
+
+    public static <T> T getBlockMeta(Location loc, String key, T def) {
+        final MetaMap blockMetaMap = getBlockMetaMap(loc, false);
+        T ret = blockMetaMap != null && !blockMetaMap.isEmpty() ? (T) blockMetaMap.get(key) : null;
+        return ret != null ? ret : def;
+    }
+
+    public static <T> T setBlockMeta(Location loc, String key, Object val) {
+        return setMetaMapValue(getBlockMetaMap(loc, true), key, val);
+    }
+
+    public static <T> T removeBlockMeta(Location loc, String key) {
+        return setMetaMapValue(getBlockMetaMap(loc, false), key, null);
+    }
+
+    //////////////////////////////////////////////////
+    //// TILE ENTITY META
+    //////////////////////////////////////////////////
+
+    public static MetaMap getTileEntityMetaMap(Location tileEntity, boolean isWrite) {
+        return MetaApiAccessor.getTileEntityMetaMap(tileEntity.getBlock().getState(), isWrite);
+    }
+    public static MetaMap getTileEntityMetaMap(BlockState tileEntity, boolean isWrite) {
+        return MetaApiAccessor.getTileEntityMetaMap(tileEntity, isWrite);
+    }
+
+    public static boolean isTileEntity(Location state) {
+        return ((CraftBlockState) state.getBlock().getState()).getTileEntity() != null;
+    }
+    public static boolean isTileEntity(BlockState state) {
+        return ((CraftBlockState) state).getTileEntity() != null;
+    }
+
+    public static boolean hasTileEntityMeta(Location tileEntity, String key) {
+        return hasTileEntityMeta(tileEntity.getBlock().getState(), key);
+    }
+
+    public static boolean hasTileEntityMeta(BlockState tileEntity, String key) {
+        final MetaMap tileEntityMetaMap = getTileEntityMetaMap(tileEntity, false);
+        return tileEntityMetaMap != null && !tileEntityMetaMap.isEmpty() && tileEntityMetaMap.containsKey(key);
+    }
+
+    public static <T> T getTileEntityMeta(Location tileEntity, String key) {
+        return getTileEntityMeta(tileEntity.getBlock().getState(), key, null);
+    }
+
+    public static <T> T getTileEntityMeta(BlockState tileEntity, String key) {
+        return getTileEntityMeta(tileEntity, key, null);
+    }
+
+    public static <T> T getTileEntityMeta(Location tileEntity, String key, T def) {
+        return getTileEntityMeta(tileEntity.getBlock().getState(), key, def);
+    }
+
+    public static <T> T getTileEntityMeta(BlockState tileEntity, String key, T def) {
+        final MetaMap tileEntityMetaMap = getTileEntityMetaMap(tileEntity, false);
+        T ret = tileEntityMetaMap != null && !tileEntityMetaMap.isEmpty() ? (T) tileEntityMetaMap.get(key) : null;
+        return ret != null ? ret : def;
+    }
+
+    public static <T> T setTileEntityMeta(Location tileEntity, String key, Object val) {
+        return setTileEntityMeta(tileEntity.getBlock().getState(), key, val);
+    }
+
+    public static <T> T setTileEntityMeta(BlockState tileEntity, String key, Object val) {
+        return setMetaMapValue(getTileEntityMetaMap(tileEntity, true), key, val);
+    }
+
+    public static <T> T removeTileEntityMeta(Location tileEntity, String key) {
+        return removeTileEntityMeta(tileEntity.getBlock().getState(), key);
+    }
+
+    public static <T> T removeTileEntityMeta(BlockState tileEntity, String key) {
+        return setMetaMapValue(getTileEntityMetaMap(tileEntity, false), key, null);
+    }
+
+    //////////////////////////////////////////////////
+    //// UTIL / CLASSES
+    //////////////////////////////////////////////////
+
+    /**
+     * General method for determining if value is null to remove it, else set.
+     * Return previous value
+     *
+     * @param map
+     * @param key
+     * @param val
+     * @param <T>
+     * @return
+     */
+    private static <T> T setMetaMapValue(MetaMap map, String key, Object val) {
+        if (map == null) {
+            return null;
+        }
+        if (val == null) {
+            return (T) map.remove(key);
+        } else {
+            return (T) map.put(key, val);
+        }
+    }
+
+    /**
+     * Checks if the passed object can be stored as Meta Data
+     * @param value
+     * @return
+     */
+    public static boolean isValidMeta(Object value) {
+        return (value instanceof String) || (value instanceof Long) ||
+            (value instanceof Integer) || (value instanceof ItemStack) ||
+            (value instanceof Float) || (value instanceof Double) ||
+            (value instanceof MetaMap) || (value instanceof MetaList);
+    }
+
+    /**
+     * A type protected array for storing meta values
+     */
+    @SuppressWarnings("PublicInnerClass")
+    public static class MetaList<T> extends ArrayList<T> {
+        public MetaList(int initialCapacity) {
+            super(initialCapacity);
+        }
+
+        public MetaList() {
+        }
+
+        public MetaList(Collection<? extends T> c) {
+            super(c);
+        }
+
+        @Override
+        public boolean add(T o) {
+            if (!isValidMeta(o)) {
+                throw new InvalidParameterException();
+            }
+            return super.add(o);
+        }
+
+        @Override
+        public T set(int index, T element) {
+            if (!isValidMeta(element)) {
+                throw new InvalidParameterException();
+            }
+            return super.set(index, element);
+        }
+
+        @Override
+        public void add(int index, T element) {
+            if (!isValidMeta(element)) {
+                throw new InvalidParameterException();
+            }
+            super.add(index, element);
+        }
+
+        public MetaList<T> clone()  {
+            return new MetaList<>(this);
+        }
+    }
+
+    /**
+     * A type protected hashmap for storing meta values
+     */
+    @SuppressWarnings("PublicInnerClass")
+    public static class MetaMap extends ForwardingMap<String, Object> {
+        private Map<String, Object> backing;
+
+        @Override
+        protected Map<String, Object> delegate() {
+            if (backing == null) {
+                backing = new HashMap<>();
+            }
+            return backing;
+        }
+
+        @Override
+        public Object put(String key, Object value) {
+            if (!isValidMeta(value)) {
+                throw new InvalidParameterException();
+            }
+            return super.put(key, value);
+        }
+
+        @Override
+        public void putAll(Map<? extends String, ?> map) {
+            for (Object value : map.values()) {
+                if (!isValidMeta(value)) {
+                    throw new InvalidParameterException();
+                }
+            }
+
+            super.putAll(map);
+        }
+
+        public MetaMap clone() {
+            MetaMap map = new MetaMap();
+            map.putAll(this);
+            return map;
+        }
+    }
+
+    @SuppressWarnings("PublicInnerClass")
+    public static final class BlockPos {
+        public final int x;
+        public final int y;
+        public final int z;
+
+        public BlockPos(int x, int y, int z) {
+            this.x = x;
+            this.y = y;
+            this.z = z;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) { return true; }
+            if (o == null || getClass() != o.getClass()) { return false; }
+
+            BlockPos blockPos = (BlockPos) o;
+
+            return (x == blockPos.x && z == blockPos.z && y == blockPos.y);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = x;
+            result = result * 73 + y;
+            result = result * 73 + z;
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "BlockPos{" +
+                "x=" + x +
+                ", y=" + y +
+                ", z=" + z +
+                '}';
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 40d4c9c..12713b0 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -112,6 +112,7 @@ public class Chunk {
 
     public org.bukkit.Chunk bukkitChunk;
     public boolean mustSave;
+    public MetaApiAccessor.ChunkMeta chunkMeta; // EMC
     // CraftBukkit end
 
     public Chunk(World world, ChunkSnapshot chunksnapshot, int i, int j) {
diff --git a/src/main/java/net/minecraft/server/ChunkRegionLoader.java b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
index be0a565..44034d0 100644
--- a/src/main/java/net/minecraft/server/ChunkRegionLoader.java
+++ b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
@@ -223,6 +223,7 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
     private void a(Chunk chunk, World world, NBTTagCompound nbttagcompound) {
         nbttagcompound.setInt("xPos", chunk.locX);
         nbttagcompound.setInt("zPos", chunk.locZ);
+        MetaApiAccessor.saveChunkMetaNbt(nbttagcompound, chunk); // EMC
         nbttagcompound.setLong("LastUpdate", world.getTime());
         nbttagcompound.setIntArray("HeightMap", chunk.r());
         nbttagcompound.setBoolean("TerrainPopulated", chunk.isDone());
@@ -329,6 +330,7 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
         Chunk chunk = new Chunk(world, i, j);
 
         chunk.a(nbttagcompound.getIntArray("HeightMap"));
+        MetaApiAccessor.loadChunkMetaNbt(nbttagcompound, chunk); // EMC
         chunk.d(nbttagcompound.getBoolean("TerrainPopulated"));
         chunk.e(nbttagcompound.getBoolean("LightPopulated"));
         chunk.c(nbttagcompound.getLong("InhabitedTime"));
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 02a1c25..32d52ee 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -43,6 +43,7 @@ import org.bukkit.plugin.PluginManager;
 public abstract class Entity implements ICommandListener {
 
     // CraftBukkit start
+    public com.empireminecraft.metaapi.MetaApi.MetaMap metaMap = null; // EMC
     private static final int CURRENT_LEVEL = 2;
     public static Random SHARED_RANDOM = new Random(); // Paper
     static boolean isLevelAtLeast(NBTTagCompound tag, int level) {
@@ -1288,6 +1289,7 @@ public abstract class Entity implements ICommandListener {
                 this.yaw = 0;
             }
 
+            MetaApiAccessor.saveEntityMeta(this, nbttagcompound); // EMC
             if (Float.isNaN(this.pitch)) {
                 this.pitch = 0;
             }
@@ -1385,6 +1387,7 @@ public abstract class Entity implements ICommandListener {
             this.motX = nbttaglist1.e(0);
             this.motY = nbttaglist1.e(1);
             this.motZ = nbttaglist1.e(2);
+            MetaApiAccessor.loadEntityMeta(this, nbttagcompound); // EMC
 
             /* CraftBukkit start - Moved section down
             if (Math.abs(this.motX) > 10.0D) {
diff --git a/src/main/java/net/minecraft/server/MetaApiAccessor.java b/src/main/java/net/minecraft/server/MetaApiAccessor.java
new file mode 100644
index 0000000..fd0eb31
--- /dev/null
+++ b/src/main/java/net/minecraft/server/MetaApiAccessor.java
@@ -0,0 +1,420 @@
+package net.minecraft.server;
+
+import com.empireminecraft.metaapi.BlockWithMetaClearedEvent;
+import net.minecraft.server.NBTBase.NBTNumber;
+import org.bukkit.craftbukkit.CraftChunk;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.regex.Pattern;
+
+import static com.empireminecraft.metaapi.MetaApi.*;
+
+public final class MetaApiAccessor {
+
+    private static final String META_MAP = "MetaMap";
+    private static final String META_TYPE = "MetaType";
+    private static final String ITEM = "Item";
+    private static final String ENTITY_META = "_EntityMeta";
+    private static final String TILEENTITY_META = "_TileEntityMeta";
+    private static final String WORLD_META = "_WorldMeta";
+    private static final String CHUNK_META_LIST = "_ChunkMetaList";
+    private static final String BLOCK_POS = "_BlockPos";
+    private static final String CHUNK_META = "_ChunkMeta";
+
+    private MetaApiAccessor() {}
+
+    public static MetaMap getWorldMetaMap(CraftWorld world) {
+        return world.getHandle().worldData.metaMap;
+    }
+
+    public static MetaMap getTileEntityMetaMap(org.bukkit.block.BlockState block, boolean isWrite) {
+        final CraftBlockState craftBlock = (CraftBlockState) block;
+        final TileEntity tileEntity = craftBlock.getTileEntity();
+        if (tileEntity == null) {
+            return null;
+        }
+        if (tileEntity.metaMap == null && isWrite) {
+            tileEntity.metaMap = new MetaMap();
+        }
+        return tileEntity.metaMap;
+    }
+
+    /**
+     * Util for getting a MetaMap by location
+     * @param loc
+     * @return
+     */
+    public static MetaMap getBlockMetaMap(org.bukkit.Location loc, boolean isWrite) {
+        return getBlockMetaMap(((CraftChunk) loc.getChunk()).getHandle(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ(), isWrite);
+    }
+
+    /**
+     * Util for getting a MetaMap for a chunk
+     * @param chunk
+     * @return
+     */
+    public static MetaMap getChunkMetaMap(org.bukkit.Chunk chunk, boolean isWrite) {
+        return getBlockMetaMap(((CraftChunk) chunk).getHandle(), 0, -1, 0, isWrite);
+    }
+
+    /**
+     * Gets the metamap for the specified coords, creating it if it doesnt exists
+     * @param handle
+     * @param x
+     * @param y
+     * @param z
+     * @return
+     */
+    private static MetaMap getBlockMetaMap(Chunk handle, int x, int y, int z, boolean isWrite) {
+        ChunkMeta meta = handle.chunkMeta;
+        if (meta == null) {
+            if (!isWrite) {
+                return null;
+            }
+            meta = handle.chunkMeta = new ChunkMeta();
+        }
+        BlockPos coords = new BlockPos(x, y, z);
+        if (!isWrite && !meta.containsKey(coords)) {
+            return null;
+        }
+        MetaMap ret = meta.get(coords);
+        if (ret == null) {
+            if (isWrite) {
+                ret = new MetaMap();
+                meta.put(coords, ret);
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Util for getting a MetaMap for an Entity
+     * @param craftentity
+     * @return
+     */
+    public static MetaMap getEntityMetaMap(org.bukkit.entity.Entity craftentity, boolean isWrite) {
+        Entity entity = ((CraftEntity) craftentity).getHandle();
+        if (entity.metaMap == null) {
+            if (isWrite) {
+                entity.metaMap = new MetaMap();
+            }
+        }
+        return entity.metaMap;
+    }
+
+    static void loadEntityMeta(Entity entity, NBTTagCompound nbt) {
+        if (nbt.hasKey(ENTITY_META)) {
+            entity.metaMap = getMetaMapFromCompound(nbt.getCompound(ENTITY_META));
+        }
+    }
+
+    static void saveEntityMeta(Entity entity, NBTTagCompound nbt) {
+        if (entity.metaMap != null && !entity.metaMap.isEmpty()) {
+            nbt.set(ENTITY_META, getNbtFromObject(entity.metaMap));
+        }
+    }
+
+    static void loadTileEntityMeta(TileEntity tileEntity, NBTTagCompound nbt) {
+        if (nbt.hasKey(TILEENTITY_META)) {
+            tileEntity.metaMap = getMetaMapFromCompound(nbt.getCompound(TILEENTITY_META));
+        }
+    }
+
+    static void saveTileEntityMeta(TileEntity tileEntity, NBTTagCompound nbt) {
+        if (tileEntity.metaMap != null && !tileEntity.metaMap.isEmpty()) {
+            nbt.set(TILEENTITY_META, getNbtFromObject(tileEntity.metaMap));
+        }
+    }
+
+    static void loadWorldMeta(WorldData world, NBTTagCompound nbt) {
+        if (nbt.hasKey(WORLD_META)) {
+            NBTTagCompound nbtmeta = nbt.getCompound(WORLD_META);
+            world.metaMap = getMetaMapFromCompound(nbtmeta);
+        }
+        if (world.metaMap == null) {
+            world.metaMap = new MetaMap();
+        }
+    }
+
+    static void saveWorldMeta(WorldData world, NBTTagCompound nbt) {
+        if (world.metaMap != null && !world.metaMap.isEmpty()) {
+            NBTBase nbtmeta = getNbtFromObject(world.metaMap);
+            nbt.set(WORLD_META, nbtmeta);
+        }
+    }
+
+    /**
+     * Saves this chunks Meta Data into NBT
+     * @param cmp
+     * @param chunk
+     */
+    static void saveChunkMetaNbt(NBTTagCompound cmp, Chunk chunk) {
+        if (chunk.chunkMeta == null || chunk.chunkMeta.isEmpty()) {
+            return;
+        }
+        NBTTagList meta = new NBTTagList();
+        for (Map.Entry<BlockPos, MetaMap> entry : chunk.chunkMeta.entrySet()) {
+            MetaMap list = entry.getValue();
+            if (!list.isEmpty()) {
+                NBTTagCompound pos = serializePos(entry.getKey());
+                NBTTagCompound metacmp = getCompoundFromMetaMap(list);
+                metacmp.set(BLOCK_POS, pos);
+                meta.add(metacmp);
+            }
+        }
+
+        if (!meta.isEmpty()) {
+            cmp.set(CHUNK_META_LIST, meta);
+        }
+    }
+
+    /**
+     * Loads this chunks Meta Data from NBT
+     * @param cmp
+     * @param chunk
+     */
+    static void loadChunkMetaNbt(NBTTagCompound cmp, Chunk chunk) {
+        ChunkMeta meta = chunk.chunkMeta;
+        if (cmp.hasKey(CHUNK_META)) {
+            NBTTagCompound chunkMeta = cmp.getCompound(CHUNK_META);
+            for (String key : chunkMeta.c()) {
+                NBTTagCompound e = chunkMeta.getCompound(key);
+                meta = loadBlockMeta(meta, deserializeOldLoc(key), e);
+            }
+        } else if (cmp.hasKey(CHUNK_META_LIST)) {
+            NBTTagList chunkMeta = cmp.getList(CHUNK_META_LIST, 10);
+            final int size = chunkMeta.size();
+            for (int i = 0; i < size; i++) {
+                NBTTagCompound e = chunkMeta.get(i);
+                if (e.hasKey(BLOCK_POS)) {
+                    NBTTagCompound blockPos = e.getCompound(BLOCK_POS);
+                    meta = loadBlockMeta(meta, getBlockPosFromCompound(blockPos), e);
+                }
+            }
+        }
+        chunk.chunkMeta = meta;
+    }
+
+    private static ChunkMeta loadBlockMeta(ChunkMeta meta, BlockPos key, NBTTagCompound e) {
+        MetaMap map = getMetaMapFromCompound(e);
+        if (map != null && !map.isEmpty()) {
+            if (meta == null) {
+                meta = new ChunkMeta();
+            }
+            meta.put(key, map);
+        }
+        return meta;
+    }
+
+    private static BlockPos getBlockPosFromCompound(NBTTagCompound pos) {
+        return new BlockPos(pos.getInt("x"), pos.getInt("y"), pos.getInt("z"));
+    }
+
+    /**
+     * Converts an Object into NBT
+     * @param value
+     * @return
+     */
+    private static NBTBase getNbtFromObject(Object value) {
+        if (value == null) {
+            return null;
+        }
+        if (value instanceof String) {
+            return new NBTTagString((String) value);
+        } else if (value instanceof ItemStack || value instanceof org.bukkit.inventory.ItemStack) {
+            ItemStack item;
+            if (value instanceof org.bukkit.inventory.ItemStack) {
+                item = CraftItemStack.asNMSCopy((org.bukkit.inventory.ItemStack) value);
+            } else {
+                item = (ItemStack) value;
+            }
+            NBTTagCompound itemnbt = new NBTTagCompound();
+            itemnbt.setString(META_TYPE, ITEM);
+            return item.save(itemnbt);
+        } else if (value instanceof Long) {
+            return new NBTTagLong((Long) value);
+        } else if (value instanceof Integer) {
+            return new NBTTagInt((Integer) value);
+        } else if (value instanceof Double) {
+            return new NBTTagDouble((Double) value);
+        } else if (value instanceof Float) {
+            return new NBTTagFloat((Float) value);
+        } else if (value instanceof MetaMap) {
+            return getCompoundFromMetaMap((MetaMap) value);
+        } else if (value instanceof MetaList) {
+            NBTTagList list = new NBTTagList();
+            for (Object obj : (Iterable<?>) value) {
+                NBTBase add = getNbtFromObject(obj);
+                if (add != null) {
+                    list.add(add);
+                }
+            }
+            return list.isEmpty() ? null : list;
+        }
+
+        return null;
+    }
+
+    /**
+     * Converts NBT into an Object
+     * @param nbt
+     * @return
+     */
+    @SuppressWarnings("RedundantCast")
+    private static Object getObjectFromNbt(NBTBase nbt) {
+        if (nbt == null) {
+            return null;
+        }
+        if (nbt instanceof NBTTagString) {
+            return nbt.a_();
+            // Redundant casts added to each number so that it will fail to compile
+            // If the method names change for each type. Do not remove.
+        } else if (nbt instanceof NBTTagInt) {
+            return (Integer) ((NBTNumber) nbt).d();
+        } else if (nbt instanceof NBTTagLong) {
+            return (Long) ((NBTNumber) nbt).c();
+        } else if (nbt instanceof NBTTagFloat) {
+            return (Float) ((NBTNumber) nbt).h();
+        } else if (nbt instanceof NBTTagDouble) {
+            return (Double) ((NBTNumber) nbt).g();
+        } else if (nbt instanceof NBTTagList) {
+            NBTTagList nbtlist = (NBTTagList) nbt;
+            if (nbtlist.isEmpty()) {
+                return null;
+            }
+            MetaList<Object> list = new MetaList<>();
+            for (int i = 0; i < nbtlist.size(); i++) {
+                final Object obj = getObjectFromNbt(nbtlist.list.get(i));
+                if (obj != null) {
+                    list.add(obj);
+                }
+            }
+            return list.isEmpty() ? null : list;
+        } else if (nbt instanceof NBTTagCompound) {
+            NBTTagCompound cmp = (NBTTagCompound) nbt.clone();
+            if (cmp.hasKey(META_TYPE)) {
+                String type = cmp.getString(META_TYPE);
+                cmp.remove(META_TYPE);
+                if (ITEM.equals(type)) {
+                    return CraftItemStack.asBukkitCopy(ItemStack.createStack(cmp));
+                } else if (META_MAP.equals(type)) {
+                    final MetaMap metaMap = getMetaMapFromCompound(cmp);
+                    if (metaMap == null || metaMap.isEmpty()) {
+                        return null;
+                    }
+                    return metaMap;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Converts a NBTTagCompound to a MetaMap
+     * @param cmp
+     * @return
+     */
+    private static MetaMap getMetaMapFromCompound(NBTTagCompound cmp) {
+        if (cmp.map.isEmpty()) {
+            return null;
+        }
+        MetaMap map = new MetaMap();
+        for (Entry<String, NBTBase> entry : cmp.map.entrySet()) {
+            final Object metaData = getObjectFromNbt(entry.getValue());
+            if (metaData != null) {
+                map.put(entry.getKey(), metaData);
+            }
+        }
+        return map;
+    }
+
+    /**
+     * Converts a MetaMap into an NBTTagCompount
+     * @param map
+     * @return
+     */
+    private static NBTTagCompound getCompoundFromMetaMap(MetaMap map) {
+        NBTTagCompound cmp = new NBTTagCompound();
+        if (map.isEmpty()) {
+            return cmp;
+        }
+        cmp.setString(META_TYPE, META_MAP);
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            NBTBase add = getNbtFromObject(entry.getValue());
+            if (add != null) {
+                cmp.set(entry.getKey(), add);
+            }
+        }
+        return cmp;
+    }
+
+    /**
+     * Translates BlockPos to a String form
+     * @param coords
+     * @return
+     */
+    private static NBTTagCompound serializePos(BlockPos coords) {
+        if (coords == null) {
+            return null;
+        }
+        NBTTagCompound cmp = new NBTTagCompound();
+        cmp.setInt("x", coords.x);
+        cmp.setInt("y", coords.y);
+        cmp.setInt("z", coords.z);
+        return cmp;
+    }
+
+    private static final Pattern COLON_SPLIT = Pattern.compile(":");
+    /**
+     * Parses a string into BlockPos
+     *
+     * @param loc
+     * @return
+     */
+    private static BlockPos deserializeOldLoc(String loc) {
+        if (loc != null)  {
+            String[] args = COLON_SPLIT.split(loc, 3);
+            if (args.length == 3) {
+                try {
+                    int x = (int) Math.floor(Float.parseFloat(args[0]));
+                    int y = (int) Math.floor(Float.parseFloat(args[1]));
+                    int z = (int) Math.floor(Float.parseFloat(args[2]));
+                    return new BlockPos(x, y, z);
+                } catch (NumberFormatException ignored) {}
+            }
+        }
+        return null;
+    }
+
+    static void processSetAir(Chunk chunk, BlockPosition pos) {
+        ChunkMeta meta = chunk.chunkMeta;
+        if (meta == null) {
+            return;
+        }
+        BlockPos coords = new BlockPos(pos.getX(), pos.getY(), pos.getZ());
+        if (meta.containsKey(coords)) {
+            new BlockWithMetaClearedEvent(chunk, pos).callEvent();
+        }
+    }
+
+    static void filterTileEntityMeta(NBTTagCompound cmp) {
+        cmp.remove(TILEENTITY_META);
+    }
+
+    static class ChunkMeta extends HashMap<BlockPos, MetaMap> {
+        ChunkMeta() {}
+
+        ChunkMeta(Map<? extends BlockPos, ? extends MetaMap> m) {
+            super(m);
+        }
+
+        public ChunkMeta clone() {
+            return new ChunkMeta(this);
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
index 4248aa1..68ba8a0 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMapChunk.java
@@ -38,6 +38,7 @@ public class PacketPlayOutMapChunk implements Packet<PacketListenerPlayOut> {
 
             if (this.e() || (i & 1 << j) != 0) {
                 NBTTagCompound nbttagcompound = tileentity.E_();
+                MetaApiAccessor.filterTileEntityMeta(nbttagcompound); // EMC
 
                 this.e.add(nbttagcompound);
             }
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutTileEntityData.java b/src/main/java/net/minecraft/server/PacketPlayOutTileEntityData.java
index b30c97e..77c7c8c 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutTileEntityData.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutTileEntityData.java
@@ -11,6 +11,7 @@ public class PacketPlayOutTileEntityData implements Packet<PacketListenerPlayOut
     public PacketPlayOutTileEntityData() {}
 
     public PacketPlayOutTileEntityData(BlockPosition blockposition, int i, NBTTagCompound nbttagcompound) {
+        MetaApiAccessor.filterTileEntityMeta(nbttagcompound); // EMC
         this.a = blockposition;
         this.b = i;
         this.c = nbttagcompound;
diff --git a/src/main/java/net/minecraft/server/TileEntity.java b/src/main/java/net/minecraft/server/TileEntity.java
index 85f97cc..6e54dad 100644
--- a/src/main/java/net/minecraft/server/TileEntity.java
+++ b/src/main/java/net/minecraft/server/TileEntity.java
@@ -50,6 +50,7 @@ public abstract class TileEntity {
     }
 
     public void a(NBTTagCompound nbttagcompound) {
+        MetaApiAccessor.loadTileEntityMeta(this, nbttagcompound); // EMC
         this.position = new BlockPosition(nbttagcompound.getInt("x"), nbttagcompound.getInt("y"), nbttagcompound.getInt("z"));
     }
 
@@ -63,6 +64,7 @@ public abstract class TileEntity {
         if (s == null) {
             throw new RuntimeException(this.getClass() + " is missing a mapping! This is a bug!");
         } else {
+            MetaApiAccessor.saveTileEntityMeta(this, nbttagcompound); // EMC
             nbttagcompound.setString("id", s);
             nbttagcompound.setInt("x", this.position.getX());
             nbttagcompound.setInt("y", this.position.getY());
@@ -257,6 +259,7 @@ public abstract class TileEntity {
         a(TileEntityCommand.class, "Control");
     }
 
+    com.empireminecraft.metaapi.MetaApi.MetaMap metaMap; // EMC
     // CraftBukkit start - add method
     public InventoryHolder getOwner() {
         if (world == null) return null;
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 0f1ee49..d585f73 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -415,6 +415,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
                 */
+                if (block == Blocks.AIR) MetaApiAccessor.processSetAir(chunk, blockposition); // EMC
 
                 // CraftBukkit start
                 if (!this.captureBlockStates) { // Don't notify clients or update physics while capturing blockstates
diff --git a/src/main/java/net/minecraft/server/WorldData.java b/src/main/java/net/minecraft/server/WorldData.java
index 6236e45..b7eca40 100644
--- a/src/main/java/net/minecraft/server/WorldData.java
+++ b/src/main/java/net/minecraft/server/WorldData.java
@@ -1,5 +1,6 @@
 package net.minecraft.server;
 
+import com.empireminecraft.metaapi.MetaApi; // EMC
 import com.google.common.collect.Maps;
 import java.util.Iterator;
 import java.util.Map;
@@ -55,6 +56,7 @@ public class WorldData {
     private final Map<DimensionManager, NBTTagCompound> N;
     private GameRules O;
     public WorldServer world; // CraftBukkit
+    public MetaApi.MetaMap metaMap = new MetaApi.MetaMap(); // EMC
 
     protected WorldData() {
         this.f = WorldType.NORMAL;
@@ -220,6 +222,7 @@ public class WorldData {
             }
         }
 
+        MetaApiAccessor.loadWorldMeta(this, nbttagcompound); // EMC
     }
 
     public WorldData(WorldSettings worldsettings, String s) {
@@ -373,7 +376,7 @@ public class WorldData {
         if (nbttagcompound1 != null) {
             nbttagcompound.set("Player", nbttagcompound1);
         }
-
+        MetaApiAccessor.saveWorldMeta(this, nbttagcompound); // EMC
     }
 
     public long getSeed() {
-- 
2.8.2

